amaranth/hdl/_ir.py

```diff
@@ -552,6 +552,8 @@ def _assign_port_names(self):
         assigned_names = {name for name, conn, dir in self.ports if name is not None}
         for name, conn, dir in self.ports:
             if name is None:
+                if conn.name == "": # Private Signal has no name so there's nothing to name this port.
+                    raise TypeError("Signals used as top-level ports must have a non-empty name")
```

Comment:
@wanda-phi Actually, do we even need this restriction?

Reply:
It doesn't break things per se, but in my experiments you do end up with a top level name which is assigned arbitrarily and so liable to change if the design changes. This seems bad for reproducibility and interoperability.

Reply:
Oh! I think that's not entirely motivating the prohibition, since it's quite obvious on examination of the exported Verilog, and we also, as far as I recall, currently don't guarantee that a certain port will have a certain name unless it's named explicitly or there are no duplicates, but the latter is another property that you can't necessarily know about the design or enforce.

So I think we should either allow that or prohibit duplicate names too. (I think allowing this is very low friction and it reduces the annoyance if you just want to look at the Verilog but don't care about synthesizing it; like in the playground for example.)

Reply:
I'm personally fine with allowing it, I was specifically asked to add it here: https://github.com/amaranth-lang/amaranth/pull/1234#discussion_r1536674341

Reply:
are you sure it doesn't break things? have you tried with *two* anonymous top-level ports? because I'm reasonably sure it'll just break, as we have no further pass that'd assign a name within NIR, and we just end up with `None` as the name and a key collision in some dict

the auto-naming you're seeing is an RTLIL backend artifact

Reply:
You are correct, with two private top-level ports, only one shows up in the Verilog. If we want to remove the restriction, I'd say teach `_add_name` to handle a name of `""` and give it an index.

Reply:
I'm fine with a restriction, I just don't want it to be artificial. Private names are going to be rarely used anyways.